### DISCUSSION TITLE
`Pack(display=NONE)` doesn't seams to be implemented on GTK

### DISCUSSION THREAD

**Discussion by ikus060** [beeware/toga#2434] [Q&A] [answered]
I continue to poke around the with toga. Sorry for all the noise I'm making, but the learning curve is kinda steep and the current state of development make it hard to know if it's a bug, a limitation of the library or my own mistake.

I'm trying to show/hide a widget from the interface by using `Pack(display=NONE)` and the feature doesn't seams to be implemented. Could anybody confirm ?

```
import toga
from toga.constants import COLUMN, ROW, NONE, HIDDEN
from toga.style.pack import Pack, PACK
import toga.colors

class TestApp(toga.App):
    def startup(self):
        self.main_window = toga.MainWindow(size=(300, 150))

        style = Pack(padding_top=24)
        substyle = Pack(padding_right=12, padding_left=12, flex=1)

        # Add the content on the main window
        self.main_window.content = toga.Box(
                    children=[
                        toga.Button(text="Toggle", on_press=self._toggle),
                        toga.TextInput(id='my_widget', placeholder="First textbox", style=Pack(visibility = HIDDEN, display=NONE, flex=1)),
                    ],
                    style=Pack(direction=COLUMN, padding=24),
                )

        # Show the main window
        self.main_window.show()

    def _toggle(self, widget, **kwargs):
        self.main_window.widgets['my_widget'].style.display = PACK


def main():
    # Application class
    #   App name and namespace
    app = TestApp("Test", "org.beeware.toga.examples.test")
    return app


if __name__ == "__main__":
    app = main()
    app.main_loop()

```

The text widget is always visible from the start.

**Comment by ikus060** [comment]
It seams there are 2 problem here. 

(1) visibility = HIDDEN is not working in GTK. call to `show_all()` update the state of all widget visibility. 

(2) display = NONE is not implemented except for CSS

**Comment by ikus060** [comment]
After some testing, I could propose a patch to fix `visibility` to avoid calling `show_all()` and make individual call to `show()`. My GTK experience is limited, so I'm not very confident, but my test tell me it's working.

```
diff --git a/gtk/src/toga_gtk/widgets/base.py b/gtk/src/toga_gtk/widgets/base.py
index d397d25ee..91fa31261 100644
--- a/gtk/src/toga_gtk/widgets/base.py
+++ b/gtk/src/toga_gtk/widgets/base.py
@@ -57,7 +57,6 @@ class Widget:
             # setting container, adding self to container.native
             self._container = container
             self._container.add(self.native)
-            self.native.show_all()
 
         for child in self.interface.children:
             child._impl.container = container
diff --git a/gtk/src/toga_gtk/window.py b/gtk/src/toga_gtk/window.py
index 0273f6242..82cab211a 100644
--- a/gtk/src/toga_gtk/window.py
+++ b/gtk/src/toga_gtk/window.py
@@ -144,7 +144,10 @@ class Window:
         app.native.add_window(self.native)
 
     def show(self):
-        self.native.show_all()
+        # Avoid calling show_all() as it change the visibility of all children.
+        self.native.show()
+        self.layout.show()
+        self.container.show()
 
     ######################################################################
     # Window content and resources
```

**Comment by ikus060** [comment, marked as answer]
I've submit a PR #2435

**Comment by freakboy3742** [comment]
> I continue to poke around the with toga. Sorry for all the noise I'm making, but the learning curve is kinda steep and the current state of development make it hard to know if it's a bug, a limitation of the library or my own mistake.

No problem about making "noise" - we never learn about bugs if people don't report them! 

As for whether something is a bug or a limitation: Toga is at v0.4.2. There's a very good chance that any problem you find is a bug or limitation. We do a lot of testing, and we have 100% test branch coverage on both the core and all the desktop and mobile backends, but even that test coverage doesn't mean things are 100% correct if the tests aren't validating the right thing.
 
> I'm trying to show/hide a widget from the interface by using `Pack(display=NONE)` and the feature doesn't seams to be implemented. Could anybody confirm ?

The example is a little odd because it includes visibility in the mix, which masks the problem - but yes, it looks like `display=NONE` isn't being honoured. I can reproduce this on both macOS and GTK. I'm not sure what has happened there - I swear it *was* working at some point, but looking at the current state of the code, there's nothing that would trigger on the value of `display`, other than CSS rendering (which is only used by the web backend).

With a few modifications to your test example (modifying the button to be a toggle, rather than just setting display=PACK, and adding a second button to toggle visibility), I *don't* see any problem with visibility, though. Both macOS and GTK seem able to toggle visibility of the text area widget. If you're seeing an issue with visibility, it's deeper than what is going on in this example.
